File: daily/30 -day finish 300 coding challenge/2020-5-14. 10 questions - Tree/98. Validate Binary Search Tree.cpp
using namespace std;
#include<vector>

struct TreeNode
{
    /* data */
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(): val(0), left(nullptr), right(nullptr){}
    TreeNode(int x): val(x), left(nullptr), right(nullptr){}
    TreeNode(int x, TreeNode * left, TreeNode *right): val(x), left(left), right(right){}
};

class Solution{
public:
    bool isValdBST(TreeNode* root, TreeNode* min=NULL, TreeNode* max=NULL){
        if(!root) return true;
        if(min != NULL && root->val <= min->val) return false;
        if(max != NULL && root->val >= max->val) return false;
        return isValdBST(root->left, min, root) && isValdBST(root->right, root, max);
    }
};

File: Daily Coding Practice/30 -day finish 300 coding challenge/2020-5-14. 10 questions - Tree/98. Validate Binary Search Tree.py
# dfs --> recursion
class Solution:
    def isValidBST(self, root):
        return self.dfs(root)
    
    def dfs(self, root, min_val = -float('inf'), max_val = float('inf')):
        if not root: return True
        if root.val <= min_val or root.val >= max_val:
            return False
        return self.root(root.left, min_val, root.val) and self.root(root.right, root.val, max_val)

# time complexity o(N)
# space complexity o(N)

#bfs --> iteration
class Solution:
    def isValidBST(self, root):
        if not root: return True
        stack = [(root, -float('inf'), float('inf'))]
        while stack:
            root, lower, upper = stack.pop()
            if not root: continue
            val = root.val
            if val <= lower or val >= upper: return False
            stack.append((root.right, val, upper))
            stack.append((root.left, lower, val))
        
        return True
# time complexity(N)
# space complexity o(N)

